File: src/bin/0224-basic-calculator.cpp
// @leetup=info id=224 lang=cpp slug=basic-calculator

#include <algorithm>
#include <cassert>
#include <cctype>
#include <iostream>
#include <map>
#include <set>
#include <stack>
#include <string>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  int calculate(std::string s) {
    int res = 0;
    std::stack<int> res_stack;
    std::stack<int> sign_stack;
    int sign = 1;
    int num = 0;
    for (char c : s) {
      if (std::isdigit(c)) {
        num = num * 10 + (c - '0');
      } else if (c == '+') {
        res += num * sign;
        num = 0;
        sign = 1;
      } else if (c == '-') {
        res += num * sign;
        num = 0;
        sign = -1;
      } else if (c == '(') {
        assert(num == 0);
        res_stack.push(res);
        sign_stack.push(sign);
        res = 0;
        sign = 1;
      } else if (c == ')') {
        res = (res + num * sign) * sign_stack.top() + res_stack.top();
        res_stack.pop();
        sign_stack.pop();

        num = 0;
        sign = 1;
      } else {
        // Skip whitespace.
      }
    }
    return res + num * sign;
  }
};

// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0224-basic-calculator.py
# @leetup=info id=224 lang=python3 slug=basic-calculator




from typing import Optional

# @leetup=code
class Solution:
    def calculate(self, s: str) -> int:
        res = 0
        res_stack = []
        sign_stack = []
        sign = 1
        num = 0
        for c in s:
            if c.isdigit():
                num = num * 10 + (ord(c) - ord("0"))
            elif c == "+":
                # ..res.. sign num +
                res += sign * num
                num = 0
                sign = 1
            elif c == "-":
                # ..res.. sign num -
                res += sign * num
                num = 0
                sign = -1
            elif c == "(":
                # res -[sign] (
                res_stack.append(res)
                sign_stack.append(sign)

                res = 0
                sign = 1
                assert num == 0
            elif c == ")":
                # 3 - (   2     -    4)
                # 3 - ( ..res.. sign num )

                # num_stack = [3]
                # sign_stack = [-1]
                # res = 2
                # sign = -1
                # num = 4

                res += sign * num
                # => res = res + sing * num = 2 + (-1 * 4) = -2

                # 3 - ( -2 )
                sign = sign_stack.pop()
                res *= sign
                # res = res * sign = -2 * -1 = 2

                # 3 + 2
                res += res_stack.pop()
                # res = res + num_stack.pop() = 2 + 3 = 5

                num = 0
                sign = 1
        return res + sign * num


# @leetup=code
